### PR TITLE
Virtual disk attach

### PR DESCRIPTION
**Pull request by PolonaM**

### CODE REVIEW

File: plugins/module_utils/disk.py [hunk 1]
```diff
@@ -72,7 +76,6 @@ def __init__(
 
     def to_hypercore(self):
         return dict(
-            uuid=self.uuid,
```
Comment (justinc1) [inline plugins/module_utils/disk.py:75]: Looks safe to me. I'm not sure if to_hypercore is ever used to POST/PATCH to server. Integration tests for vm, vm_disk do pass?

File: tests/unit/plugins/module_utils/test_virtual_disk.py [hunk 2]
```diff
@@ -790,3 +790,21 @@ def test_send_delete_request_virtual_disk(
...
+            name="",
+            block_size="",
+            size="",
+            replication_factor="",
+        )
+        virtual_disk.attach_to_vm(rest_client, None)
+
+        rest_client.create_record.assert_called_with(
+            endpoint="/rest/v1/VirtualDisk/asd-123123/attach",
+            payload=None,
```
Comment (justinc1) [inline tests/unit/plugins/module_utils/test_virtual_disk.py:808]: Can you add test where payload will not be None?

File: plugins/modules/virtual_disk_attach.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+  - module: scale_computing.hypercore.virtual_disk_info
+  - module: scale_computing.hypercore.vm_disk
+options:
+  name:
+    description:
+      - Virtual disk name that we want to attach.
+      - Used as a unique identifier of an uploaded virtual disk.  # should we use UUID??
+    type: str
+    required: true
+  block_device: # in vm_disk this is named "items", should we use disk?
```
Comment (justinc1) [inline plugins/modules/virtual_disk_attach.py:35]: I would like to have similar naming as vm and vm_disk modules. vm_disk: `items` does not fit (plural), and `item` is also ugly. vm: `disks` is plural => you win, `disk` looks like best name to me.

**Review by justinc1 — CHANGES_REQUESTED** [review]
smaller changes. thank you.

File: plugins/modules/virtual_disk_attach.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+        required: true
+      disk_slot:
+        description:
+          - Virtual slot the drive will occupy.
+          - If not defined, the next empty slot will be occupied.
+        type: int
+      size:
+        description:
+          - Logical size of the block device in bytes.
+          - Should be greater or equal than the size of virtul_disk. If smaller then capacity of the new block device
```
[the remark this answers is not in the document]
Comment (PolonaM) [reply plugins/modules/virtual_disk_attach.py:55]: Updated

File: plugins/modules/virtual_disk_attach.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+  - scale_computing.hypercore.vm_name
+seealso:
+  - module: scale_computing.hypercore.virtual_disk
+  - module: scale_computing.hypercore.virtual_disk_info
+  - module: scale_computing.hypercore.vm_disk
+options:
+  name:
+    description:
+      - Virtual disk name that we want to attach.
+      - Used as a unique identifier of an uploaded virtual disk.  # should we use UUID??
```
[the remark this answers is not in the document]
Comment (PolonaM) [reply plugins/modules/virtual_disk_attach.py:32]: Comment removed

File: plugins/modules/virtual_disk_attach.py [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+      tiering_priority_factor: 8
+      read_only: true
+      regenerate_disk_id: false
+  register: block_device
+"""
+
+RETURN = r"""
+record:
+  description:
+    - Created block device.
```
[the remark this answers is not in the document]
Comment (PolonaM) [reply plugins/modules/virtual_disk_attach.py:115]: Explained!

File: plugins/modules/virtual_disk_attach.py [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+) -> Dict[Any, Any]:
+    payload = {}
+    payload["options"] = dict(
+        regenerateDiskID=module.params["block_device"]["regenerate_disk_id"],
+        readOnly=module.params["block_device"]["read_only"],
+    )
+    payload["template"] = Disk.from_ansible(  # type: ignore
+        module.params["block_device"]
+    ).post_and_patch_payload(vm)
+    payload["template"].pop("readOnly")
```
[the remark this answers is not in the document]
Comment (PolonaM) [reply plugins/modules/virtual_disk_attach.py:202]: It was confusing for me to, but it is ok :) The payload looks like this: 
{
  "options": {
    "regenerateDiskID": true,
    "readOnly": false
  },
  "template": {
    "virDomainUUID": "string",
    "type": "IDE_DISK",
    "cacheMode": "NONE",
    "capacity": 0,
    "shareUUID": "string",
    "path": "string",
    "slot": -1,
    "name": "string",
    "disableSnapshotting": true,
    "tieringPriorityFactor": 0
  }
}
readOnly is under "options" and not under "template" which is actually disk

File: tests/integration/targets/virtual_disk_attach/tasks/01_supported.yml [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+      - block_device is succeeded
+      - block_device is changed
+      - >
+        block_device.record.keys() | sort == ['cache_mode', 'disable_snapshotting', 'disk_slot', 'iso_name',
+        'mount_points', 'read_only', 'size', 'tiering_priority_factor', 'type', 'uuid', 'vm_uuid']
+      - block_device.record.type == "virtio_disk"
+      - block_device.record.disk_slot == 3
+      - block_device.record.size == 104857600
+      - block_device.record.iso_name == ""
+      - block_device.record.tiering_priority_factor == 8 # THIS DOESN'T ALWAYS WORK - RETURNS 0
```
[the remark this answers is not in the document]
Comment (PolonaM) [reply tests/integration/targets/virtual_disk_attach/tasks/01_supported.yml:108]: Then I should just comment out the assertion?

File: tests/integration/targets/virtual_disk_attach/tasks/01_supported.yml [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+      - block_device.record.cache_mode == "writethrough"
+      - block_device.record.disable_snapshotting == true
+      - block_device.record.read_only == true
+      - block_device.record.vm_uuid == vm.records.0.uuid
+      - block_device.diff.after == block_device.record
+      - block_device.diff.before == block_device.record
+
+- name: Attach virtual disk with only type defined (test default values)
```
[the remark this answers is not in the document]
Comment (PolonaM) [reply tests/integration/targets/virtual_disk_attach/tasks/01_supported.yml:158]: True, it will create disk in the next empty slot if the slot is not defined. Slot is not required in payload, thus I didn't require it in the module, but it makes sense to be required. I updated the module.

File: plugins/module_utils/disk.py (hunk 1 above)
Comment (PolonaM) [reply plugins/module_utils/disk.py:75]: Yes, vm and vm_disk pass

File: plugins/modules/virtual_disk_attach.py (hunk 3 above)
Comment (PolonaM) [reply plugins/modules/virtual_disk_attach.py:35]: Updated to "disk"

File: plugins/modules/virtual_disk_attach.py [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+from ..module_utils.client import Client
+from ..module_utils.rest_client import RestClient
+from ..module_utils.virtual_disk import VirtualDisk
+from ..module_utils.disk import Disk
+from ..module_utils.task_tag import TaskTag
+from ..module_utils.vm import VM
+from ..module_utils.hypercore_version import HyperCoreVersion
+
+
+HYPERCORE_VERSION_REQUIREMENTS = ">=9.2.10"  # WHICH VERSION?
```
[the remark this answers is not in the document]
Comment (PolonaM) [reply plugins/modules/virtual_disk_attach.py:181]: Nice suggestion, I updated the modules

File: tests/unit/plugins/module_utils/test_virtual_disk.py (hunk 2 above)
Comment (PolonaM) [reply tests/unit/plugins/module_utils/test_virtual_disk.py:808]: Added

File: plugins/modules/virtual_disk_attach.py [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+    description:
+      - The block device that will be created when selected virtual disk is attached to selected virtual machine.
+    type: dict
+    suboptions:
+      type:
+        description:
+          - The bus type the block device will use.
+          - If I(type=ide_cdrom), I(iso_name) is also required. Se documentation of I(iso_name) for more details.
+        type: str
+        choices: [ ide_cdrom, virtio_disk, ide_disk, scsi_disk, ide_floppy, nvram ]
```
[the remark this answers is not in the document]
Comment (PolonaM) [reply plugins/modules/virtual_disk_attach.py]: You are right, updated as suggested

File: plugins/modules/virtual_disk_attach.py [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+    size:
+      description: Logical size of the device in bytes, and can be increased on update or clone
+      type: int
+      sample: 81001000100
+    disk_slot:
+      description: Virtual slot the drive will occupy
+      type: int
+      sample: 0
+    iso_name:
+      description: Name of the virtual storage device
```
[the remark this answers is not in the document]
Comment (PolonaM) [reply plugins/modules/virtual_disk_attach.py:138]: Option iso_name deleted

File: changelogs/fragments/virtual_disk_attach_module.yml [hunk 13]
```diff
@@ -0,0 +1,3 @@
+---
+major_changes:
+  - Added virtual_disk_attach module. (https://github.com/ScaleComputing/HyperCoreAnsibleCollection/pull/221)
```
Comment (justinc1) [inline changelogs/fragments/virtual_disk_attach_module.yml:3]: typo, 212

File: examples/virtual_disk_attach.yml [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+        that:
+          - vm_info_result.records | length == 1
+          - virtual_disk_info_result.records | length == 1
+
+    # ------------------------------------------------------
+    - name: Attach virtual disk {{ virtual_disk_name }} to VM {{ vm_name }} {{ vm_disk_type }}:{{ vm_disk_slot }}
+      scale_computing.hypercore.virtual_disk_attach:
+        name: "{{ virtual_disk_name }}"
+        vm_name: "{{ vm_name }}"
+        block_device:
```
Comment (justinc1) [inline examples/virtual_disk_attach.yml:36]: `disk` or `block_device`?

File: changelogs/fragments/virtual_disk_attach_module.yml (hunk 13 above)
Comment (PolonaM) [reply changelogs/fragments/virtual_disk_attach_module.yml:3]: Thanks for noticing, corrected

**Review by justinc1 — APPROVED** [review]
thank you
